design/20190708.certificate-request-crd.md

```diff
@@ -174,6 +188,190 @@ type ObjectReference struct {
+rejected by the API server.
+
+Setting the Approved or Denied conditions are restricted by the approver having
+sufficient RBAC permissions. These permissions are based upon the request
+itself- specifically the request's IssuerRef:
+
+```yaml
+apiGroups: ["cert-manager.io"]
+resources: ["signers"]
+verbs: ["approve"]
+resourceNames:
+ # namesapced signers
+ - "<signer-resource-name>.<signer-group>/<signer-namespace>.<signer-name>"
+ # cluster scoped signers
+ - "<signer-resource-name>.<signer-group>/<signer-name>"
+ # all signers of this resource name
+ - "<signer-resource-name>.<signer-group>/*"
+```
+
+An example ClusterRole that would grant the permissions to set the Approve and
+Denied conditions of CertificateRequests that reference the cluster scoped
+`myissuers` external issuer, in the group `my-example.io`, with the name `myapp`:
+
+```
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: my-example-io-my-issuer-myapp-approver
+rules:
+  - apiGroups: ["cert-manager.io"]
+    resources: ["signers"]
+    verbs: ["approve"]
+    resourceNames: ["myissuers.my-example.io/myapp"]
+```
+
+These permissions are enforced using a
+[SubjectAccessReview](https://github.com/kubernetes/kubernetes/blob/b11d0fbdd58394a62622787b38e98a620df82750/pkg/apis/authorization/types.go#L27)
+that is performed in the cert-manager Webhook component when a user attempts to
+set the Approve or Denied condition. The user of the approver is tested along
+with the IssuerRef of the CertificateRequest.
+
+If the approver does not have sufficient permissions defined above to set the
+Approved or Denied conditions, the request will be rejected by the cert-manager
+validating admission webhook.
+
+###### Scope
+The RBAC permissions *must* be granted at the cluster scope (i.e.
+ClusterRoleBinding). This matches the same scoping behaviour as
+[certificates.k8s.io
+CertificateSigningRequests](https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/#authorization).
+
+Namespaced signers are represented by a namespaced resource using the syntax of
+`<signer-resource-name>.<signer-group>/<signer-namespace>.<signer-name>`
+
+###### API Group
+The apiGroup must *always* be `cert-manager.io`, as all CertificateRequests are
+in the `cert-manager.io` group. Even though some signers may themselves not
+belong to the `cert-manager.io` group, the approve operation does.
+
+###### Resource
+The resource must *always* be `signers`. This matches the same resource as
+[required by certificates.k8s.io
+CertificateSigningRequests](https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/#authorization).
+
+###### Verbs
+The verb must *always* be `approve`. This verb grants the approver the
+permissions to set *both* Approved and Denied conditions.
+
+###### Resource Names
+The resource name is made up of 3 parts: the signer resource name, signer group, and the
+signer name.
```

Comment:
It would be worth using the typography `apiGroups`, `resources`, `verbs` and `resourceNames` in these 4 sections to help the reader understand that these subsections refer to the above RBAC role.

I suggest using a table:

```suggestion
|      Field      |                                                      Description                                                       |
|-----------------|------------------------------------------------------------------------------------------------------------------------|
| `apiGroups`     | Must be `cert-manager.io`. This is required for the `approve` verb, even if an external issuers has a different group. |
| `resources`     | Must be `signers`. This is similar to [CertificateSigningRequests][csr].                                               |
| `verbs`         | Must be `signers`. Gives permission to set the `Approved` and `Denied` conditions.                                     |
| `resourceNames` | Uses a custom syntax described below.                                                                                  |


[csr]: https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/#authorization

###### resourceNames custom syntax
```

Reply:
Thanks, I will update the names. Not keen on tables at it makes it more difficult to read in a text editor.